data/finset/basic.lean

```diff
@@ -0,0 +1,704 @@
+/-
+Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Leonardo de Moura, Jeremy Avigad
+
+Finite sets.
+-/
+
+import data.list.set data.list.perm tactic.finish
+open list subtype nat
+
+def nodup_list (A : Type) := {l : list A // nodup l}
+
+variable {α : Type}
+
+def to_nodup_list_of_nodup {l : list α} (n : nodup l) : nodup_list α := 
+⟨l, n⟩
+
+def to_nodup_list [decidable_eq α] (l : list α) : nodup_list α :=
+@to_nodup_list_of_nodup α (erase_dup l) (nodup_erase_dup l)
+
+private def eqv (l₁ l₂ : nodup_list α) := 
+perm l₁.1 l₂.1
+
+local infix ~ := eqv
+
+private def eqv.refl (l : nodup_list α) : l ~ l :=
+perm.refl _
+
+private def eqv.symm {l₁ l₂ : nodup_list α} : l₁ ~ l₂ → l₂ ~ l₁ :=
+perm.symm
+
+private def eqv.trans {l₁ l₂ l₃ : nodup_list α} : l₁ ~ l₂ → l₂ ~ l₃ → l₁ ~ l₃ :=
+perm.trans
+
+instance finset.nodup_list_setoid  (α : Type) : setoid (nodup_list α) :=
+setoid.mk (@eqv α) (mk_equivalence (@eqv α) (@eqv.refl α) (@eqv.symm α) (@eqv.trans α))
+
+def finset (α : Type) : Type :=
+quotient (finset.nodup_list_setoid α)
+
+namespace finset
+
+def to_finset_of_nodup (l : list α) (n : nodup l) : finset α :=
+⟦to_nodup_list_of_nodup n⟧
+
+def to_finset [decidable_eq α] (l : list α) : finset α :=
+⟦to_nodup_list l⟧
+
+lemma to_finset_eq_of_nodup [decidable_eq α] {l : list α} (n : nodup l) :
+  to_finset_of_nodup l n = to_finset l :=
+have P : to_nodup_list_of_nodup n = to_nodup_list l, from
+  begin
+    dsimp [to_nodup_list], have eq : erase_dup l = l, 
+    {apply erase_dup_eq_of_nodup n},
+    {simp [eq]} -- rw [eq] doesn't work
+  end,
+quotient.sound (eq.subst P (setoid.refl _))
+
+instance has_decidable_eq  [decidable_eq α] : decidable_eq (finset α) :=
+λ s₁ s₂, quotient.rec_on_subsingleton₂ s₁ s₂
+  (λ l₁ l₂,
+     match perm.decidable_perm l₁.1 l₂.1 with
+     | decidable.is_true e := decidable.is_true (quot.sound e)
+     | decidable.is_false n := decidable.is_false (λ e : ⟦l₁⟧ = ⟦l₂⟧, absurd (quotient.exact e) n)
+end)
+
+def mem (a : α) (s : finset α) : Prop :=
+quot.lift_on s (λ l, a ∈ l.1)
+ (λ l₁ l₂ (e : l₁ ~ l₂), propext (iff.intro
+   (λ ainl₁, perm.mem_of_perm e ainl₁)
+   (λ ainl₂, perm.mem_of_perm (perm.symm e) ainl₂)))
+
+-- instance : has_mem α (finset α) := ⟨mem⟩  -- this causes a failure of type class inference
```

Comment:
this is strange, where does the failure happen?

Reply:
Thanks for the comments! Others are fixed except for this one, and the following type class instances that depend on it. It seems that lean cannot figure out that the `⟦l⟧` in `mem_of_mem_list` is a finset when using type class instances. Replacing the `∈` by `mem` fixes the problem, but then this doesn't seem to be as convenient as using notations. Similar things happen when adding type class instances for other constants. I have to write everything explicitly like `@singleton _ (finset α) _ _ a` to make lean aware of the correct types.

Reply:
For singleton it should be enough to write `{a} : finset α`, for most type classes, i.e. intersection, union, subset, it is enough if the elaborator can determine from one side what the type is (i.e. in `a ∪ ∅` it is enough to know the type of `a`). For a statement like `∅ ⊆ {a}` you need to annotate the type.

The problem with the syntax you introduce is that we are now get into overloading, and I'm never sure how well this is actually handled. All the constants you introduce (`∅`, `⊆`, `∪`, `∈`, `∩`) are used for sets, and are provided as type class instances.

Did you try to use `finset` together with `set`? I guess one gets a couple of problems...

Reply:
Yes, annotating types works for singleton. However, lean is still complaining in some other places, for example, the following theorem doesn't work. 

`
theorem mem_singleton_iff (x a : α) : x ∈ ({a} : finset α) ↔ (x = a) :=
by rewrite [(mem_insert_eq x a ∅), mem_empty_eq, or_false]
`

rewrite tactic failed, did not find instance of the pattern in the target expression
  x ∈ insert a ∅
state:
α : Type u,
h : decidable_eq α,
x a : α
⊢ x ∈ {a} ↔ x = a

I agree that overloading notations cause problems... Maybe the safest way for now is to just define these type class instances but call the constants by name instead of notations?

Reply:
I agree with @johoelzl that we should use the type classes here, even if it means adding extra information in a few places, because the overloading will cause problems.

For example, the problem with `a ∈ ⟦l⟧` is that the type of the right-hand side is only definitionally equal to `finset`. You can fix that one by writing `a ∈ @id (finset α) ⟦l⟧` or `a ∈ (show finset α, from ⟦l⟧)` or `@has_mem.mem _ (finset α) _ a ⟦l⟧` or `((∈) : α → finset α → Prop) a ⟦l⟧`. I wish there were a nicer solution...
￼  ￼  ￼￼

Reply:
Thanks for the solution. I guess I will just write `mem a ⟦l⟧` then... But for the above one I am still not sure what was going on. It checks if the goal is `x ∈ insert a ∅ ↔ (x = a)`, but fails if it is `x ∈ ({a} : finset α) ↔ (x = a)`, given that we have introduced the type class instance `has_insert`.

Reply:
I have found part of the problem: the rewriter doesn't recognize that `insert` is the same as `has_insert.insert`. If you change the theorems that mention `insert` to mention `have_insert.insert`, the proof works. An even better solution is to declare
```
attribute [reducible] insert
```
at the top of the file. (Maybe this should be done already in the core library, but for now we can do it here and anywhere else that uses `insert`.)

@johoelzl What do you think?

Reply:
Either `[reducible] insert`, or we ask Leo if it makes sense to use the `insert` constant for the `{_}` and `{_, ..., _}` syntax.